From 4618585843beea7cf157ad4585dbe6ce4052e94b Mon Sep 17 00:00:00 2001
From: JBirdVegas <jbirdvegas@gmail.com>
Date: Sun, 4 Nov 2012 23:30:52 -0500
Subject: [PATCH] Mute aapt

---
 tools/aapt/ResourceTable.cpp | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/tools/aapt/ResourceTable.cpp b/tools/aapt/ResourceTable.cpp
index 99f9d2f..476007e 100644
--- a/tools/aapt/ResourceTable.cpp
+++ b/tools/aapt/ResourceTable.cpp
@@ -2612,6 +2612,7 @@ ResourceTable::validateLocalizations(void)
          nameIter++) {
         const set<String8>& configSet = nameIter->second;   // naming convenience
 
+#ifdef SHOW_DEFAULT_TRANSLATION_WARNINGS
         // Look for strings with no default localization
         if (configSet.count(defaultLocale) == 0) {
             fprintf(stderr, "aapt: warning: string '%s' has no default translation in %s; found:",
@@ -2626,7 +2627,8 @@ ResourceTable::validateLocalizations(void)
                 err = BAD_VALUE;
             }
         }
-
+#endif
+#ifdef SHOW_LOCALIZATION_WARNINGS
         // Check that all requested localizations are present for this string
         if (mBundle->getConfigurations() != NULL && mBundle->getRequireLocalization()) {
             const char* allConfigs = mBundle->getConfigurations();
@@ -2665,6 +2667,7 @@ ResourceTable::validateLocalizations(void)
                 }
            } while (comma != NULL);
         }
+#endif
     }
 
     return err;
